RedBookPlayer.Models/Hardware/Karaoke/KaraokeDisplay.cs

```diff
@@ -0,0 +1,367 @@
+        private void SetBorderColor(BorderPreset borderPreset)
+        {
+            if(borderPreset == null)
+                return;
+
+            for(int x = 0; x < 3; x++)
+            for(int y = 0; y < 216; y++)
+            {
+                this.DisplayData[x,y] = borderPreset.Color;
+            }
+
+            for(int x = 297; x < 300; x++)
+            for(int y = 0; y < 216; y++)
+            {
+                this.DisplayData[x,y] = borderPreset.Color;
+            }
+
+            for(int x = 0; x < 300; x++)
+            for(int y = 0; y < 6; y++)
+            {
+                this.DisplayData[x,y] = borderPreset.Color;
+            }
+
+            for(int x = 0; x < 300; x++)
+            for(int y = 210; y < 216; y++)
+            {
+                this.DisplayData[x,y] = borderPreset.Color;
+            }
+        }
+    
+        /// <summary>
+        /// Load a block of pixels with a certain pattern
+        /// </summary>
+        /// <param name="tileBlock">TileBlock with the pattern data</param>
+        /// <param name="xor">
+        /// If true, the color values are combined with the color values
+        /// that are already onscreen using the XOR operator
+        /// </param>
+        private void LoadTileBlock(TileBlock tileBlock, bool xor)
+        {
+            if(tileBlock == null)
+                return;
+
+            // Extract out the "bitmap" into a byte pattern
+            byte[,] pattern = new byte[12,6];
+            for(int i = 0; i < tileBlock.TilePixels.Length; i++)
+            {
+                byte b = tileBlock.TilePixels[i];
+                pattern[i,0] = (byte)(b & (1 << 0));
+                pattern[i,1] = (byte)(b & (1 << 1));
+                pattern[i,2] = (byte)(b & (1 << 2));
+                pattern[i,3] = (byte)(b & (1 << 3));
+                pattern[i,4] = (byte)(b & (1 << 4));
+                pattern[i,5] = (byte)(b & (1 << 5));
+            }
+
+            // Now load the bitmap starting in the correct place
+            for(int x = 0; x < 12; x++)
+            for(int y = 0; y < 6; y++)
+            {
+                int adjustedX = x + tileBlock.Column;
+                int adjustedY = y + tileBlock.Row;
+                byte colorIndex = pattern[x,y] == 0 ? tileBlock.Color0 : tileBlock.Color1;
+
+                if(xor)
+                    this.DisplayData[adjustedX, adjustedY] = (byte)(colorIndex ^ this.DisplayData[adjustedX, adjustedY]);
+                else
+                    this.DisplayData[adjustedX, adjustedY] = colorIndex;
+            }
+        }
+
+        /// <summary>
+        /// Scroll the display according to the instruction
+        /// </summary>
+        /// <param name="scroll">Scroll with the new data</param>
+        /// <param name="copy">True if data wraps around on scroll, false if filled by a solid color</param>
+        private void ScrollDisplay(Scroll scroll, bool copy)
+        {
+            if(scroll == null)
```

Comment:
What happens if scroll contains negatives due to errors?

Reply:
I can add the secondary checks for negatives here.